### DISCUSSION TITLE
How to extract all data from jwt callback

### DISCUSSION THREAD

**Discussion by roxennnn** [nextauthjs/next-auth#4296] [Help]
### Question 💬

I have been struggling to extract all the fields from a NextAuth.js OAuth provider. Specifically, in the case of a Google provider, I was able to extract these fields:

```json
{
  "expires": "",
  "token": {
    "name": "",
    "email": "",
    "picture": "",
    "sub": "",
    "iat": 0,
    "exp": 0,
    "jti": ""
  }
}
```

However, when logging inside the `jwt` function, I can see that there are many more fields (e.g., the `profile` field):

```json
{
  "profile": {
    "iss": "",
    "azp": "",
    "aud": "",
    "sub": "",
    "hd": "",
    "email": "",
    "email_verified": true,
    "at_hash": "",
    "name": "",
    "picture": "",
    "given_name": "",
    "family_name": "",
    "locale": "",
    "iat": 0,
    "exp": 0
  }
},
```

**What I would like to get access in my `sign-in` page are all the data fields from the `jwt` callback parameters:**

```ts
jwt: (params: {
  token: JWT;
  user?: User | undefined;
  account?: Account | undefined;
  profile?: Profile | undefined;
  isNewUser?: boolean | undefined;
}) => Awaitable<JWT>
```

Logging around, it also seems that the `jwt` callback is called twice for some reason and, in the second call, there are not all the fields. According to the [documentation](https://next-auth.js.org/configuration/callbacks#jwt-callback):

> The arguments user, account, profile and isNewUser are only passed the first time this callback is called on a new session, after the user signs in. In subsequent calls, only token will be available.

I guess that this explains why the second time `jwt` is called carries less data, but still it does not explain why it gets called twice.

### How to reproduce ☕️

I have already read and tried basically every tutorial and documentation on the [NextAuth.js](https://next-auth.js.org/) website. For clarity's sake, I report hereafter the pieces of code relevant:

**\_app.tsx**:

```tsx
function MyApp({ Component, pageProps: { session, ...pageProps } }: AppProps) {
  return (
    <SessionProvider session={session}>
      <Component {...pageProps} />
    </SessionProvider>
  );
}
```

**\[...nextauth\].ts**

```ts
export default NextAuth({
  providers: [
    GoogleProvider({
      clientId: '',
      clientSecret: '',
    }),
  ],
  secret: 'secret token',
  pages: {
    signIn: '/sign-in',
  },
  session: {
    strategy: 'jwt',
  },
  //Callback here
  callbacks: {
    jwt: (params: {
      token: JWT;
      user?: User | undefined;
      account?: Account | undefined;
      profile?: Profile | undefined;
      isNewUser?: boolean | undefined;
    }): Awaitable<JWT> => {
      const { token, user, account, profile, isNewUser } = params;
      if (account?.accessToken) {
        token.accessToken = account.accessToken;
      }

      console.log({ ...token, user, account, profile, isNewUser });

      return { ...token, user, account, profile, isNewUser };
    },
    async session({ session, token, user }) {
      return { ...session, user, token };
    },
  },
});
```

**./pages/sign-in.tsx**

```tsx
...
  const { data: session } = useSession();

  useEffect(() => {
    console.log(session);
  }, [session]);
...
```

Is there anything I am missing to solve my problem?

### Contributing 🙌🏽

No, I am afraid I cannot help regarding this

**Comment by ephraimduncan** [comment]
From the documentation:

> The session callback is called whenever a session is checked. By default, only a subset of the token is returned for increased security. If you want to make something available you added to the token through the `jwt()` callback, you have to explicitly forward it here to make it available to the client. e.g. `getSession()`, `useSession()`, `/api/auth/session`

Add a `session` callback to the callbacks then return an object with the `session` and the `token`. You can access the `accessToken` from the client-side using `getSession()`, `useSession()` or `/api/auth/session`

```js
async session({ session, token }) {
  return { ...session, ...token }
},
```

**Comment by EyadAlsaba** [comment]
Hey @roxennnn 
I'm having a problem with extracting any field from the " token ,account or user " 
can you please share the code you have written to make this work?
I really appreciate any help you can provide.

**Reply by roxennnn** [reply]
Hey, I don't really remember what I did specifically. What I recall though is that I created an endpoint where I stored the whole jwt content and then I could request it. This was just a workaround though.

Haven't you tried the solution from the previous answer?

**Reply by EyadAlsaba** [reply]
Thanks for the response, 
I have tried the solution from this discussion, which is the only resource I've found!
My problem is more in jwt callback! somehow the function does not return anything!!! however, I'm able to see what it holds in the console!!! very strange behaviour 
according to the docs, this jwt callback is called whenever you hit a certain point like (useSession,getSession, etc), but the session object is just undefined!!!
using JS the logic works as intended, with TS something is missing and I don't see where and why!!!
for the sake of clarity, I will post my code, maybe if you see it, you will be able to think of something! 
----
export const authOptions: NextAuthOptions = {
  providers: [
    SpotifyProvider({
      clientId: `${process.env.SPOTIFY_CLIENT_ID}`,
      clientSecret: `${process.env.SPOTIFY_CLIENT_SECRET}`,
      authorization: SPOTIFY_AUTHORIZATION_URL,
    }),
  ],
  secret: process.env.NEXTAUTH_SECRET,
  pages: {
    signIn: "/auth/login",
  },
  session: {
    strategy: "jwt",
  },
  callbacks: {
    jwt: (params: {
      token: JWT;
      user?: User | undefined;
      account?: Account | undefined | null;
    }): Awaitable<JWT> => {
      const { token, user, account } = params;
      console.log(
        "token =>",
        token,
        "\n user =>",
        user,
        "\n account =>",
        account,
      );
      if (account && user && account.expires_at) {
        console.log("if block fire");
        spotify_Web_API.setAccessToken(account.access_token);
        spotify_Web_API.setRefreshToken(account.refresh_token);
        return {
          accessToken: account.access_token,
          accessTokenExpires: Date.now() + account.expires_at * 1000,
          refreshToken: account.refresh_token,
          user,
        };
      }
      // Return previous token if the access token has not expired yet
      if (Date.now() < account?.expires_at! * 1000) {
        return token;
      }
      return refreshAccessToken(account!);
    },
    session: (params: { session: Session; user: User; token: JWT }) => {
      const { session, token, user } = params;
      console.log(
        "session token =>",
        token,
        "\n session user",
        user,
        "\n session",
        session.user,
      );
      return { ...session, user, ...token };
    },
  },
  debug: false,
};
export default NextAuth(authOptions);

**Reply by roxennnn** [reply]
Could you please format your code? It would improve clarity

**Reply by ephraimduncan** [reply]
```ts
export const authOptions: NextAuthOptions = {
  providers: [
    SpotifyProvider({
      clientId: process.env.SPOTIFY_CLIENT_ID,
      clientSecret: process.env.SPOTIFY_CLIENT_SECRET,
      authorization: SPOTIFY_AUTHORIZATION_URL,
    }),
  ],
  secret: process.env.NEXTAUTH_SECRET,
  pages: {
    signIn: "/auth/login",
  },
  session: {
    strategy: "jwt",
  },
  callbacks: {
    jwt: (params: {
      token: JWT;
      user?: User | undefined;
      account?: Account | undefined | null;
    }): Awaitable => {
      const { token, user, account } = params;
      console.log("token =>", token, "\n user =>", user, "\n account =>", account);
      if (account && user && account.expires_at) {
        console.log("if block fire");
        spotify_Web_API.setAccessToken(account.access_token);
        spotify_Web_API.setRefreshToken(account.refresh_token);
        return {
          accessToken: account.access_token,
          accessTokenExpires: Date.now() + account.expires_at * 1000,
          refreshToken: account.refresh_token,
          user,
        };
      }
      // Return previous token if the access token has not expired yet
      if (Date.now() < account?.expires_at! * 1000) {
        return token;
      }
      return refreshAccessToken(account!);
    },
    session: (params: { session: Session; user: User; token: JWT }) => {
      const { session, token, user } = params;
      console.log("session token =>", token, "\n session user", user, "\n session", session.user);
      return { ...session, user, ...token };
    },
  },
  debug: false,
};
export default NextAuth(authOptions);

```

**Reply by EyadAlsaba** [reply]
@dephraiim thank you
I just saw the comment! 
please let me know if you have noticed anything!
